File: parsing/backtrack/backtrack/BacktrackParser.cpp
#include "BacktrackParser.h"

t::BacktrackParser::BacktrackParser(BacktrackLexer* le) :lexer(le) {
	markers = new std::vector<int>();
	lookahead = new std::vector<Token*>();
	sync(1);
}

void t::BacktrackParser::mark() {
	markers->push_back(p);
	
}

void t::BacktrackParser::release() {
	p = markers->back();
	markers->pop_back();
}

t::Token* t::BacktrackParser::LT(int i) {
	sync(i);
	return lookahead->at(p+i-1);
}

int t::BacktrackParser::LA(int i) {
	return LT(i)->type;
}

void t::BacktrackParser::sync(int i) {
	int t1 = p + i - 1;
	int t2 = lookahead->size() - 1;
	if (t1 > t2) {
		int n = t1 - t2;
		for (int i = 1; i <= n; i++) {
			lookahead->push_back(lexer->nextToken());
		}
	}
}

void t::BacktrackParser::consume() {
	p++;
	// ·ÇÍÆ¶Ï×´Ì¬£¬
	if (p == lookahead->size()&&markers->size()<=0) {
		p = 0;
		lookahead->clear();
	}
	sync(1);
}

void t::BacktrackParser::stat() {
	if (specluate_rule_1()) {
		assign();
		match(BacktrackLexer::EOF_TYPE);
		std::cout << "like [a,b]=[c,d]" << std::endl;
	}
}
bool t::BacktrackParser::speculate_rule_2() {
	bool success = true;

	return success;
}

bool t::BacktrackParser::specluate_rule_1() {
	bool success = true;
	mark();
	success = assign() && match(BacktrackLexer::EOF_TYPE);
	release();
	return success;
}
bool t::BacktrackParser::list() {
	return match(BacktrackLexer::LBRACK) && elements() && match(BacktrackLexer::RBRACK);
}

bool t::BacktrackParser::elements() {
	bool success = true;
	success &= element();
	while (LA(1) == BacktrackLexer::COMMA) {
		success&=match(BacktrackLexer::COMMA);
		success&=element();
	}
	return success;
}

bool t::BacktrackParser::element() {

	if (LA(1) == BacktrackLexer::NAME&&LA(2) == BacktrackLexer::EQUALS) {
		return match(BacktrackLexer::NAME) && match(BacktrackLexer::EQUALS) && match(BacktrackLexer::NAME);
	}
	else if (LA(1) == BacktrackLexer::NAME) {
		return match(BacktrackLexer::NAME);
	}
	else if (LA(1) == BacktrackLexer::LBRACK) {
		return list();
	}
	else {
		std::cout << "invalid token " << LT(1)->toString() << std::endl;
		return false;
	}

}
bool t::BacktrackParser::assign() {
	return list() && match(BacktrackLexer::EQUALS) && list();
}

bool t::BacktrackParser::match(int i) {
	if (LA(1) == i) {
		consume();
		return true;
	}
	else {
		std::cout << "expecting " << BacktrackLexer::tokenNames[i] << "; found " << LT(i)->value << std::endl;
		return false;
	}
}

File: java/scala-compiler/compiler/src/main/java/com/moon/memorize/BacktrackParser.java
package com.moon.memorize;

import java.util.*;

/**
 * Created by lin on 2/11/17.
 */
public class BacktrackParser {
    public static final int FAILED=-1;
    BacktrackLexer input;
    List<Integer> markers;
    List<Token> lookahead;
    Map<Integer,Integer> listMemo=new HashMap<Integer,Integer>();
    int mark=0;

    public BacktrackParser(BacktrackLexer input) {
        this.input = input;
        markers=new ArrayList<>();
        lookahead=new ArrayList<>();
        sync(1);
    }

    public void stat(){
        if(speculate_stat_alt_1()){
            list();
            match(BacktrackLexer.EOF_TYPE);
        }else if(speculate_stat_alt_2()){
            System.out.println("predict alternative 2");
            assign();
            match(BacktrackLexer.EOF_TYPE);
        }
    }
    public void sync(int i){
        if(mark+i-1>(lookahead.size()-1)){
            int n=(mark+i-1)-(lookahead.size()-1);
            fill(n);
        }
    }

    public void fill(int n){
        for(int i=1;i<=n;i++){
            lookahead.add(input.nextToken());
        }
    }

    public Token token(int i){
        sync(i);
        return lookahead.get(mark+i-1);
    }

    public int type(int i){
        return token(i).getType();
    }

    public void match(int x){
        if(type(1)==x)consume();
        else throw new MismatchedTokenException("excepting "+BacktrackLexer.tokenNames[x]+"; found "+type(1));
    }

    public void consume(){
        mark++;
        if(mark==lookahead.size()&&!(markers.size()>0)){
            mark=0;
            lookahead.clear();
            listMemo.clear();
        }
    }

    public void mark(){
        markers.add(mark);
    }

    public void release(){
        mark=markers.get(markers.size()-1);
        markers.clear();
    }

    public boolean alreadyParsedRule(Map<Integer,Integer> memorization){
        Integer memoi=memorization.get(mark);
        if(Objects.isNull(memoi))
            return false;
        System.out.println("parsed list before at index "+mark+"; skip ahead to token index "+memoi+":"+lookahead.get(memoi).getValue());
        if(memoi==FAILED)
            throw new PreviousParsedException();
        mark=memoi;
        return true;
    }

    public void memorize(Map<Integer,Integer> memorization,int startTokenIndex,boolean failed){
        int stopTokenIndex=failed?FAILED:mark;
        memorization.put(startTokenIndex,stopTokenIndex);
    }

    public boolean speculate_stat_alt_1(){
        System.out.println("attempt alternative 1");
        boolean success=true;
        mark();
        try{
            list();
            match(BacktrackLexer.EOF_TYPE);
        }catch(RuntimeException e){
            success=false;
        }
        release();
        return success;
    }

    public boolean speculate_stat_alt_2(){
        System.out.println("attempt alternative 2");
        boolean success=true;
        mark();
        try{
            assign();
            match(BacktrackLexer.EOF_TYPE);
        }catch (RuntimeException e){
            success=false;
        }
        release();
        return success;
    }
    public void assign(){
        list();
        match(BacktrackLexer.EQUALS);
        list();
    }
    public void list(){
        boolean failed=false;
        int startTokenIndex=mark;
        if(markers.size()>0&&alreadyParsedRule(listMemo))
            return;
        try{
            _list();
        }catch(RuntimeException e){
            failed=true;
            throw e;
        }finally {
            if(markers.size()>0)
                memorize(listMemo,startTokenIndex,failed);
        }

    }

    public void _list(){
        System.out.println("parsed list rule at token index: "+mark);
        match(BacktrackLexer.LBRACK);
        elements();
        match(BacktrackLexer.RBRACK);
    }

    public void elements(){
        element();
        while(type(1)==BacktrackLexer.COMMA){
            match(BacktrackLexer.COMMA);
            element();
        }
    }

    public void element(){
        if(type(1)==BacktrackLexer.NAME&&type(2)==BacktrackLexer.EQUALS){
            match(BacktrackLexer.NAME);
            match(BacktrackLexer.EQUALS);
            match(BacktrackLexer.NAME);
        }else if(type(1)==BacktrackLexer.NAME){
            match(BacktrackLexer.NAME);
        }else if(type(1)==BacktrackLexer.EQUALS){
            match(BacktrackLexer.EQUALS);
        }else throw new NoViableException("expecting element found "+type(1));
    }
}
